File: frontend/src/lib/predicates.ts
export interface PredicateParamDef {
  key: string;
  label: string;
  type: "number" | "string";
  default?: string | number;
}

export interface PredicateDef {
  id: string;
  label: string;
  description: string;
  circuit: string;
  discloses: string[];
  hides: string[];
  params: PredicateParamDef[];
}

export const PREDICATES: PredicateDef[] = [
  {
    id: "cgpa_gte",
    label: "CGPA at or above a threshold",
    description: "Prove CGPA ≥ a cutoff without revealing the exact score.",
    circuit: "ClaimProver",
    discloses: ["Whether CGPA meets the threshold", "That the credential is not revoked"],
    hides: ["Exact CGPA", "Degree title", "Graduation date"],
    params: [{ key: "threshold", label: "Minimum CGPA", type: "number", default: 7 }],
  },
  {
    id: "degree_eq",
    label: "Holds a specific degree",
    description: "Prove the committed degree title matches.",
    circuit: "commitment-bound",
    discloses: ["Whether the degree title matches", "That the credential is not revoked"],
    hides: ["CGPA", "Other degree details"],
    params: [
      {
        key: "degree",
        label: "Degree title",
        type: "string",
        default: "Bachelor of Computer Engineering",
      },
    ],
  },
  {
    id: "year_range",
    label: "Graduated in a year range",
    description: "Prove issue year falls within a range.",
    circuit: "commitment-bound",
    discloses: ["Whether graduation year is in range", "That the credential is not revoked"],
    hides: ["Exact date", "CGPA"],
    params: [
      { key: "from_year", label: "From year", type: "number", default: 2020 },
      { key: "to_year", label: "To year", type: "number", default: 2026 },
    ],
  },
  {
    id: "graduated",
    label: "Has graduated",
    description: "Boolean: an active, anchored credential exists.",
    circuit: "NonRevocation",
    discloses: ["That a valid degree credential exists and is not revoked"],
    hides: ["CGPA", "Degree title", "Graduation date"],
    params: [],
  },
  {
    id: "issuer_set",
    label: "Issued by an approved university",
    description: "Prove the issuer DID is in a requested allow-list.",
    circuit: "IssuerMembership",
    discloses: ["That the issuer is in the verifier's allow-list"],
    hides: ["All academic attributes"],
    params: [
      {
        key: "issuer_dids",
        label: "Allowed issuer DIDs (comma-separated)",
        type: "string",
      },
    ],
  },
];

export function getPredicate(id?: string | null): PredicateDef {
  return PREDICATES.find((p) => p.id === id) ?? PREDICATES[0];
}

export function evaluateLocalPredicate(
  predicate: string,
  params: Record<string, unknown> | undefined,
  subject: { degree?: string; cgpa?: number | string; date?: string } | undefined,
  issuerDid: string | undefined,
  status: string
): { pass: boolean; reason: string } {
  const p = params ?? {};
  if (status && status.toLowerCase() !== "active") {
    return { pass: false, reason: "revoked" };
  }
  if (predicate === "cgpa_gte") {
    const threshold = Number(p.threshold ?? 0);
    const cgpa = Number(subject?.cgpa ?? 0);
    return cgpa >= threshold
      ? { pass: true, reason: "pass" }
      : { pass: false, reason: "threshold_miss" };
  }
  if (predicate === "degree_eq") {
    const wanted = String(p.degree ?? "").trim().toLowerCase();
    const got = String(subject?.degree ?? "").trim().toLowerCase();
    return wanted && got === wanted
      ? { pass: true, reason: "pass" }
      : { pass: false, reason: "degree_mismatch" };
  }
  if (predicate === "year_range") {
    const from = Number(p.from_year ?? 0);
    const to = Number(p.to_year ?? 9999);
    const year = Number(String(subject?.date ?? "").slice(0, 4));
    return year >= from && year <= to
      ? { pass: true, reason: "pass" }
      : { pass: false, reason: "year_out_of_range" };
  }
  if (predicate === "issuer_set") {
    const allowed = String(p.issuer_dids ?? "")
      .split(",")
      .map((s) => s.trim().toLowerCase())
      .filter(Boolean);
    if (!allowed.length) return { pass: true, reason: "pass" };
    return allowed.includes((issuerDid ?? "").toLowerCase())
      ? { pass: true, reason: "pass" }
      : { pass: false, reason: "issuer_not_in_set" };
  }
  return { pass: true, reason: "pass" };
}

export function failReasonLabel(reason?: string | null): string {
  switch (reason) {
    case "threshold_miss":
      return "The CGPA threshold was not met";
    case "revoked":
      return "This credential has been revoked";
    case "degree_mismatch":
      return "The degree title does not match";
    case "year_out_of_range":
      return "Graduation year is outside the requested range";
    case "issuer_not_in_set":
      return "The issuer is not on the approved list";
    case "expired":
      return "This request has expired";
    default:
      return reason ? reason.replace(/_/g, " ") : "The proof did not pass";
  }
}

File: backend/app/core/predicates.py
"""Predicate helpers over already-committed credential fields."""
from datetime import datetime
from typing import Any, Dict, Optional, Tuple


PREDICATE_CATALOG = [
    {
        "id": "cgpa_gte",
        "label": "CGPA at or above a threshold",
        "description": "Prove CGPA ≥ a cutoff without revealing the exact score.",
        "circuit": "ClaimProver",
        "discloses": ["Whether CGPA meets the threshold", "That the credential is not revoked"],
        "hides": ["Exact CGPA", "Degree title", "Graduation date"],
        "params": [{"key": "threshold", "label": "Minimum CGPA", "type": "number", "default": 7.0}],
    },
    {
        "id": "degree_eq",
        "label": "Holds a specific degree",
        "description": "Prove the committed degree title matches (equality).",
        "circuit": "commitment-bound",
        "discloses": ["Whether the degree title matches", "That the credential is not revoked"],
        "hides": ["CGPA", "Other degree details"],
        "params": [{"key": "degree", "label": "Degree title", "type": "string", "default": "Bachelor of Computer Engineering"}],
    },
    {
        "id": "year_range",
        "label": "Graduated in a year range",
        "description": "Prove issue year falls within a range.",
        "circuit": "commitment-bound",
        "discloses": ["Whether graduation year is in range", "That the credential is not revoked"],
        "hides": ["Exact date", "CGPA"],
        "params": [
            {"key": "from_year", "label": "From year", "type": "number", "default": 2020},
            {"key": "to_year", "label": "To year", "type": "number", "default": 2026},
        ],
    },
    {
        "id": "graduated",
        "label": "Has graduated",
        "description": "Boolean: an active, anchored credential exists.",
        "circuit": "NonRevocation",
        "discloses": ["That a valid degree credential exists and is not revoked"],
        "hides": ["CGPA", "Degree title", "Graduation date"],
        "params": [],
    },
    {
        "id": "issuer_set",
        "label": "Issued by an approved university",
        "description": "Prove the issuer DID is in a requested set (IssuerMembership).",
        "circuit": "IssuerMembership",
        "discloses": ["That the issuer is in the verifier's allow-list"],
        "hides": ["All academic attributes"],
        "params": [{"key": "issuer_dids", "label": "Allowed issuer DIDs (comma-separated)", "type": "string"}],
    },
]


def evaluate_predicate(
    predicate: str,
    params: Optional[Dict[str, Any]],
    subject: Optional[Dict[str, Any]],
    issuer_did: Optional[str],
    status: str,
) -> Tuple[bool, str]:
    params = params or {}
    subject = subject or {}
    if status and status.lower() not in ("active", "issued"):
        return False, "revoked"

    if predicate == "cgpa_gte":
        threshold = float(params.get("threshold", 0) or 0)
        cgpa = float(subject.get("cgpa") or 0)
        if cgpa >= threshold:
            return True, "pass"
        return False, "threshold_miss"

    if predicate == "degree_eq":
        wanted = str(params.get("degree") or "").strip().lower()
        got = str(subject.get("degree") or subject.get("degreeName") or "").strip().lower()
        if wanted and got == wanted:
            return True, "pass"
        return False, "degree_mismatch"

    if predicate == "year_range":
        from_year = int(params.get("from_year") or 0)
        to_year = int(params.get("to_year") or 9999)
        date = str(subject.get("date") or "")
        year = None
        if date:
            try:
                year = int(date[:4])
            except ValueError:
                year = None
        if year is None:
            year = datetime.utcnow().year
        if from_year <= year <= to_year:
            return True, "pass"
        return False, "year_out_of_range"

    if predicate == "graduated":
        return True, "pass"

    if predicate == "issuer_set":
        raw = str(params.get("issuer_dids") or "")
        allowed = [p.strip().lower() for p in raw.split(",") if p.strip()]
        if not allowed:
            return True, "pass"
        if (issuer_did or "").lower() in allowed:
            return True, "pass"
        return False, "issuer_not_in_set"

    return True, "pass"


def catalog_entry(predicate_id: str) -> Optional[Dict[str, Any]]:
    for item in PREDICATE_CATALOG:
        if item["id"] == predicate_id:
            return item
    return PREDICATE_CATALOG[0]
